dali/imgcodec/image_decoder.cc

```diff
@@ -0,0 +1,772 @@
+
+  return params.size();
+}
+
+
+DecodeResult ImageDecoder::Decode(DecodeContext ctx,
+                                  SampleView<CPUBackend> out,
+                                  ImageSource *in,
+                                  DecodeParams opts,
+                                  const ROI &roi) {
+  return ScheduleDecode(ctx, out, in, opts, roi).get_one(0);
+}
+
+DecodeResult ImageDecoder::Decode(DecodeContext ctx,
+                                  SampleView<GPUBackend> out,
+                                  ImageSource *in,
+                                  DecodeParams opts,
+                                  const ROI &roi) {
+  return ScheduleDecode(ctx, out, in, opts, roi).get_one(0);
+}
+
+
+/**
+ * @brief Decodes a single image to device buffers
+ */
+std::vector<DecodeResult> ImageDecoder::Decode(DecodeContext ctx,
+                                               span<SampleView<CPUBackend>> out,
+                                               cspan<ImageSource *> in,
+                                               DecodeParams opts,
+                                               cspan<ROI> rois) {
+  return ScheduleDecode(ctx, out, in, opts, rois).get_all_copy();
+}
+
+
+/**
+ * @brief Decodes a single image to device buffers
+ */
+std::vector<DecodeResult> ImageDecoder::Decode(DecodeContext ctx,
+                                               span<SampleView<GPUBackend>> out,
+                                               cspan<ImageSource *> in,
+                                               DecodeParams opts,
+                                               cspan<ROI> rois) {
+  return ScheduleDecode(ctx, out, in, opts, rois).get_all_copy();
+}
+
+
+
+FutureDecodeResults ImageDecoder::ScheduleDecode(DecodeContext ctx,
+                                                 SampleView<CPUBackend> out,
+                                                 ImageSource *in,
+                                                 DecodeParams opts,
+                                                 const ROI &roi) {
+  return ScheduleDecode(ctx, make_span(&out, 1), make_span(&in, 1), opts, make_span(&roi, 1));
+}
+
+FutureDecodeResults ImageDecoder::ScheduleDecode(DecodeContext ctx,
+                                                 SampleView<GPUBackend> out,
+                                                 ImageSource *in,
+                                                 DecodeParams opts,
+                                                 const ROI &roi) {
+  return ScheduleDecode(ctx, make_span(&out, 1), make_span(&in, 1), opts, make_span(&roi, 1));
+}
+
+FutureDecodeResults ImageDecoder::ScheduleDecode(DecodeContext ctx,
+                                                span<SampleView<CPUBackend>> out,
+                                                cspan<ImageSource *> in,
+                                                DecodeParams opts,
+                                                cspan<ROI> rois) {
```

Comment:
We ignore the `rois` parameter here (except for asserting its correctness :P), probably we'd like to copy it to `work->rois`

Reply:
Fixed in ScheduledWork::init